src/test/java/games/strategy/test/EqualsPredicate.java

```diff
@@ -0,0 +1,72 @@
+package games.strategy.test;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+final class EqualsPredicate {
+  private final EqualityComparatorRegistry equalityComparatorRegistry;
+
+  EqualsPredicate(final EqualityComparatorRegistry equalityComparatorRegistry) {
+    this.equalityComparatorRegistry = equalityComparatorRegistry;
+  }
+
+  boolean test(final @Nullable Object o1, final @Nullable Object o2) {
+    return new Context().equals(o1, o2);
+  }
+
+  private final class Context implements EqualityComparator.Context {
+    private final Set<ActiveComparison> activeComparisons = new HashSet<>();
+
+    @Override
+    public boolean equals(final @Nullable Object o1, final @Nullable Object o2) {
+      if (o1 == o2) {
+        return true;
+      } else if (o1 == null) {
+        return o2 == null;
+      } else if (o2 == null) {
+        return false; // o1 cannot be null here
+      }
+
+      final ActiveComparison activeComparison = new ActiveComparison(o1, o2);
+      if (activeComparisons.contains(activeComparison)) {
+        return true; // short-circuit active comparisons; they will eventually be resolved
+      }
```

Comment:
This is how the equality test handles circular references.  If a comparison is currently active higher up the call stack, it just bails out here and keeps the test going by returning `true`.  The initial comparison will eventually complete with the correct result (see the circular reference tests for scenarios that demonstrate this).

Returning `true` here may smell a bit.  I considered having a three-state return value: `EQUAL`, `NOT_EQUAL`, and `PENDING`, where I would return `PENDING` in this case.  However, it made the actual `EqualityComparator` implementations much more complex.  So I preferred having a small smell in one location rather than placing a burden on devs who will be implementing many more custom comparators.